### PR TITLE
[documentation update] update doc for pad

### PR DESCRIPTION
**Pull request by nswamy**
screen shot 2017-04-21 at 2 14 30 pm

### CODE REVIEW

**Comment by nswamy** [comment]
@mli, @zackchase , @madjam, @Roshrini , @indhub , @jiajiechen 
Please review

File: src/operator/pad.cc [hunk 1]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
 DMLC_REGISTER_PARAMETER(PadParam);
 
 MXNET_REGISTER_OP_PROPERTY(Pad, PadProp)
-.describe(R"code(Pad an array.
+.describe(R"code(Pads an input array with a constant or edge values of the array.
 
-Only supports 4-D and 5-D input array.
+.. note:: `Pad` is deprecated. Use `pad` instead.
+
+.. note:: Current Implementation only supports 4-D or 5-D input arrays with padding applied only on the lower
```
Comment (madjam) [inline src/operator/pad.cc]: Implementation -> implementation
"4D or 5D" -> "4D and 5D"
"on the lower" -> "to the lower"

File: src/operator/pad-inl.h [hunk 2]
```diff
@@ -88,8 +86,8 @@ class PadOp : public Operator {
           out_data[pad_enum::kOut].get<xpu, 5, DType>(s);
       pad_image(out, data, param_.pad_width, param_.mode, constant_value);
     } else {
-      LOG(FATAL) << "Only 4d or 5d input tensors with padding applied to "
-                    "dimensions > 1 is currently implemented.";
+      LOG(FATAL) << "Current Implementation only supports 4-D or 5-D input arrays with padding applied "
```
Comment (madjam) [inline src/operator/pad-inl.h]: Implementation -> implementation
"4D or 5D" -> "4D and 5D"
"on the lower" -> "to the lower"

File: src/operator/pad-inl.h [hunk 3]
```diff
@@ -88,8 +86,8 @@ class PadOp : public Operator {
...
+                    "only on the lower 2 or 3 dimensions(expects higher 2 dimensions in pad_width to be zero).";
```
Comment (madjam) [inline src/operator/pad-inl.h]: space before (

File: src/operator/pad.cc [hunk 4]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
...
+   2 or 3 dimensions(expects higher 2 dimensions in `pad_width` to be zero).
```
Comment (madjam) [inline src/operator/pad.cc]: space before (
"expects higher 2 dimensions in `pad_width` to be zero" -> "higher 2 dimensions in `pad_width` must be zero"

File: src/operator/pad-inl.h (hunk 3 above)
Comment (madjam) [inline src/operator/pad-inl.h]: "expects higher 2 dimensions in `pad_width` to be zero" -> "higher 2 dimensions in `pad_width` must be zero"

File: src/operator/pad.cc [hunk 5]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
...
+
+This operation pads an input array with either `constant_value` or edge values
+along each axis of the input array according to `pad_width`.
```
Comment (madjam) [inline src/operator/pad.cc]: Suggest rewording this as:

.. along each axis of the input array. The amount of padding is specified by `pad_width`.

File: src/operator/pad.cc [hunk 6]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
...
+
+`pad_width` is a tuple of integer padding widths for each axis of the format
+``(before_1, after_1, ... , before_N, after_N)``. The `pad_width` should be of length ``2*N``
+where ``N`` is the number of dimensions of the array.
+
+For dimension ``N`` of the input array, ``before_N`` and ``after_N`` indicates how many values
+to add before and after the contents of the array along the dimension ``N``.
```
Comment (madjam) [inline src/operator/pad.cc]: contents -> elements

Comment (madjam) [inline src/operator/pad.cc]: drop "the" before dimension.

File: src/operator/pad.cc [hunk 7]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
...
+The widths of the higher two dimensions ``before_1``, ``after_1``, ``before_2``,
+``after_2`` should be 0.
```
Comment (madjam) [inline src/operator/pad.cc]: should -> must

File: src/operator/pad.cc [hunk 8]
```diff
@@ -400,12 +400,65 @@ Operator *PadProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
...
+          [ 10.  11.  12.]]]
+
+
+        [[[ 11.  12.  13.]
+          [ 14.  15.  16.]]
+
+         [[ 17.  18.  19.]
+          [ 20.  21.  22.]]]]
+
+   pad(x,mode="edge", pad_width=(0,0,0,0,1,1,1,1)) =
```
Comment (madjam) [inline src/operator/pad.cc]: Please add an example for constant

File: src/operator/pad-inl.h [hunk 9]
```diff
@@ -37,21 +37,18 @@ struct PadParam : public dmlc::Parameter<PadParam> {
...
+            "Padding type to use."
+            " \"constant\" pads with `constant_value` and"
+            " \"edge\" pads using the edge values of the input array.");
 
     DMLC_DECLARE_FIELD(pad_width).describe(
-        "A tuple of padding widths of length 2*r, where r is the rank of the "
-        "input tensor, specifying number of values padded to the edges of each "
-        "axis. (before_1, after_1, ... , before_N, after_N) unique pad widths "
-        "for each axis. Equivalent to pad_width in numpy.pad, but flattened.");
+        "Number of values padded to the edges of each axis. "
```
Comment (zackchase) [inline src/operator/pad-inl.h]: I believe this is not factually correct. The padding width is not the "number of values". For exmple if you padded a matrix 10 by 10 with (3, 3, 3, 3) That would correspond to 156 *values* (in the sense of the number of components in the padded region. Better to say:

"Widths of the padding regions applied to the edges of each axis"

File: src/operator/pad-inl.h (hunk 2 above)
Comment (zackchase) [inline src/operator/pad-inl.h]: Agree with @madjam 👍

File: src/operator/pad-inl.h (hunk 3 above)
Comment (zackchase) [inline src/operator/pad-inl.h]: Could you be more specific than "lower 2 or 3 dimensions". I'm not sure it's clear to the reader. Also I think this is may muddle the use of "dimensions" vs "axis" in the way that we're trying to avoid.

You could say "with padding applied only on axes 1, 2, and 3. Expects axes 4 and 5 in `pad_width` to be zero"?. 

Also, note pad_width should be in backticks.

File: src/operator/pad.cc (hunk 1 above)
Comment (zackchase) [inline src/operator/pad.cc]: agreed 👍

File: src/operator/pad.cc (hunk 4 above)
Comment (zackchase) [inline src/operator/pad.cc]: See my above comment

File: src/operator/pad-inl.h (hunk 9 above)
Comment (nswamy) [inline src/operator/pad-inl.h]: This is similar to numpy [pad_width](https://docs.scipy.org/doc/numpy/reference/generated/numpy.pad.html). The description was taken from there. Changing to what you suggest which removes ambiguity

File: src/operator/pad-inl.h (hunk 2 above)
Comment (nswamy) [inline src/operator/pad-inl.h]: changed.

File: src/operator/pad-inl.h (hunk 3 above)
Comment (nswamy) [inline src/operator/pad-inl.h]: changed the message. 
this is a log message so backticks won't matter here.

Comment (nswamy) [inline src/operator/pad-inl.h]: modified message per Zack's advise.

File: src/operator/pad.cc (hunk 1 above)
Comment (nswamy) [inline src/operator/pad.cc]: changed.

File: src/operator/pad.cc (hunk 4 above)
Comment (nswamy) [inline src/operator/pad.cc]: changed based on previous comment.

File: src/operator/pad.cc (hunk 5 above)
Comment (nswamy) [inline src/operator/pad.cc]: changed.

File: src/operator/pad.cc (hunk 8 above)
Comment (nswamy) [inline src/operator/pad.cc]: done

File: src/operator/pad.cc (hunk 7 above)
Comment (nswamy) [inline src/operator/pad.cc]: done

File: src/operator/pad.cc (hunk 6 above)
Comment (nswamy) [inline src/operator/pad.cc]: done

**Comment by zackchase** [comment]
What's the status on this? Why hasn't it been merged @piiswrong ?
